miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultLibraryService.java

```diff
@@ -0,0 +1,550 @@
+    return libraryDao.getLibraryStrategyTypeByName(name);
+  }
+
+  @Override
+  public Collection<LibraryStrategyType> getAllLibraryStrategyTypes() throws IOException {
+    return libraryDao.listAllLibraryStrategyTypes();
+  }
+
+  @Override
+  public Collection<QcType> getAllLibraryQcTypes() throws IOException {
+    return libraryQcDao.listAllLibraryQcTypes();
+  }
+
+  @Override
+  public void addNote(Library library, Note note) throws IOException {
+    Library managed = libraryDao.get(library.getId());
+    authorizationManager.throwIfNotWritable(managed);
+    note.setCreationDate(new Date());
+    note.setOwner(authorizationManager.getCurrentUser());
+    managed.addNote(note);
+    save(managed);
+  }
+
+  @Override
+  public void deleteNote(Library library, Long noteId) throws IOException {
+    if (noteId == null || noteId.equals(Note.UNSAVED_ID)) {
+      throw new IllegalArgumentException("Cannot delete an unsaved Note");
+    }
+    Library managed = libraryDao.get(library.getId());
+    authorizationManager.throwIfNotWritable(managed);
+    Note deleteNote = null;
+    for (Note note : managed.getNotes()) {
+      if (note.getNoteId().equals(noteId)) {
+        deleteNote = note;
+        break;
+      }
+    }
+    if (deleteNote == null) {
+      throw new IOException("Note " + noteId + " not found for Library  " + library.getId());
+    }
+    authorizationManager.throwIfNonAdminOrMatchingOwner(deleteNote.getOwner());
+    managed.getNotes().remove(deleteNote);
+    libraryDao.save(managed);
+  }
+
+  @Override
+  public void addQc(Library library, LibraryQC qc) throws IOException {
+    Library managed = libraryDao.get(library.getId());
+    authorizationManager.throwIfNotWritable(managed);
+    qc.setQcCreator(authorizationManager.getCurrentUsername());
+    try {
+      qc.setLibrary(managed);
+      managed.addQc(qc);
+    } catch (MalformedLibraryException e) {
+      throw new IOException("Malformed Library " + managed.getName() + " in the database...", e);
+    } catch (MalformedLibraryQcException e) {
+      throw new IOException("Malformed Library QC");
+    }
+    libraryQcDao.save(qc);
+    libraryDao.save(library);
+  }
+
+  @Override
+  public void deleteQc(Library library, Long qcId) throws IOException {
+    if (qcId == null || qcId.equals(AbstractLibraryQC.UNSAVED_ID)) {
+      throw new IllegalArgumentException("Cannot delete an unsaved Library QC");
+    }
+    Library managed = libraryDao.get(library.getId());
+    authorizationManager.throwIfNotWritable(managed);
+    LibraryQC deleteQc = null;
+    for (LibraryQC qc : managed.getLibraryQCs()) {
+      if (qc.getId() == qcId) {
+        deleteQc = qc;
+        break;
+      }
+    }
+    if (deleteQc == null) throw new IOException("QC " + qcId + " not found for Library " + library.getId());
+    authorizationManager.throwIfNonAdminOrMatchingOwner(securityManager.getUserByLoginName(deleteQc.getQcCreator()));
```

Comment:
It's not a problem with qcCreator being null. It's a problem with qcCreator being a String that is not an existing MISO username. If `securityManager.getUserByLoginName(deleteQc.getQcCreator())` returns null, then `authorizationManager.throwIfNonAdminOrMatchingOwner(securityManager.getUserByLoginName(deleteQc.getQcCreator()))` will throw a NPE. The correct behaviour would be for `throwIfNonAdminOrMatchingOwner` to only check for admin status if the "MatchingOwner" parameter is null